src/java/org/apache/cassandra/metrics/DecayingEstimatedHistogramReservoir.java

```diff
@@ -855,4 +855,319 @@ public String toString()
+                {
+                    decaying[i] = buckets[i];
+                    estimated[i] = buckets[size + i];
+                }
+                for (BucketsThreadLocal local : locals)
+                {
+                    DecayingArray decayingLoc = local.decayingRef.get();
+                    long[] estimatedLoc = local.estimated;
+                    for (int i = 0; i < size; i++)
+                    {
+                        decaying[i] += decayingLoc.data[i];
+                        estimated[i] += estimatedLoc[i];
+                    }
+                }
+            } while (!stampedLock.validate(stamp));
+
+            return new DecayingEstimatedArray(decaying, estimated, resultLandmark);
+        }
+
+        private void flushPendingExclusive()
+        {
+            assert stampedLock.isWriteLocked();
+            DecayingArray arr;
+            while ((arr = decayingPending.poll()) != null)
+            {
+                // We need to flush the values to the decaying buckets only, which is a half of the buckets array.
+                for (int i = 0; i < arr.data.length; i++)
+                    buckets[i] += arr.data[i];
+            }
+        }
+    }
+
+    private static class DecayingEstimatedArray
+    {
+        private final long[] decaying;
+        private final long[] estimated;
+        private final long decayLandmark;
+
+        public DecayingEstimatedArray(long[] decaying, long[] estimated, long decayLandmark)
+        {
+            this.decaying = decaying;
+            this.estimated = estimated;
+            this.decayLandmark = decayLandmark;
+        }
+
+        public long[] estimated()
+        {
+            return estimated;
+        }
+
+        public long[] decaying()
+        {
+            return decaying;
+        }
+
+        public long landmark()
+        {
+            return decayLandmark;
+        }
+    }
+
+    /**
+     * This class is used to store the decaying buckets in a thread local variable along with the landmark.
+     * No concurrency issues are expected here, as the thread local is only used by one thread at a time.
+     */
+    private static class DecayingArray
+    {
+        private final long[] data;
+        private final long decayLandmark;
+        /**
+         * As SampledClock is used to register the last time the decay weight was sampled,
+         * and the precision of the clock is not guaranteed to be nanoseconds (approximately 2ms),
+         * we can avoid calculating the decay weight for every sample and instead use the last calculated weight.
+         */
+        private long lastSampledClock;
+        private long lastDecayedWeight;
+
+        public DecayingArray(int size, long decayLandmark)
+        {
+            this.data = new long[size];
+            this.decayLandmark = decayLandmark;
+        }
+
+        public void update(int index, long now)
+        {
+            if (lastSampledClock != now)
```

Comment:
forwardDecayWeight actually uses seconds granularity, so we can reuse the value for much longer..